catalog/security/banner-marking/src/main/java/org/codice/alliance/security/banner/marking/BannerCommonMarkingExtractor.java

```diff
@@ -13,77 +13,119 @@
+                    .put(Security.DISSEMINATION_CONTROLS, this::processDissem)
+                    // At the present moment, we are going to be filling the security.owner-producer and
+                    // security.classification-system fields with the same content, the trigraph/tetragraph
+                    // of the originating country/organization. That may diverge in the future
+                    .put(Security.OWNER_PRODUCER, this::processOwnerProducer)
+                    .put(Security.CLASSIFICATION_SYSTEM, this::processClassSystem)
+                    .build();
 
-    public static final String SECURITY_DISSEMINATION_CONTROLS = "security.dissemination-controls";
-
-    public static final String SECURITY_OWNER_PRODUCER = "security.owner-producer";
+    public BannerCommonMarkingExtractor() {
+        setAttProcessors(securityMap);
+    }
 
-    public static final String SECURITY_CLASSIFICATION_SYSTEM = "security.classification-system";
+    private void checkSecurityAttribute(String name, Serializable oldVal, Serializable newVal) {
+        if (!isSecurityAttribute(name) || StringUtils.isEmpty((String) oldVal)) {
+            return;
+        }
 
-    public BannerCommonMarkingExtractor() {
-        Map<String, BiFunction<Metacard, BannerMarkings, Attribute>> tempMap = new HashMap<>();
+        if (!newVal.equals(oldVal)) {
+            String errorMessage = String.format(
+                    "Extracted security attribute %s from the banner markings"
+                            + " does not match value pre-existing on the metacard. The"
+                            + " metacard will not be created.",
+                    name);
+            INGEST_LOGGER.error(errorMessage);
+            throw new RuntimeException(errorMessage);
+        }
+    }
 
-        tempMap.put(SECURITY_CLASSIFICATION, this::processClassMarking);
-        tempMap.put(SECURITY_RELEASABILITY, this::processReleasability);
-        tempMap.put(SECURITY_CODEWORDS, this::processCodewords);
-        tempMap.put(SECURITY_DISSEMINATION_CONTROLS, this::processDissem);
+    private void checkSecurityAttributes(String name, List<Serializable> oldVals,
+            List<Serializable> newVals) {
+        if (!isSecurityAttribute(name) || CollectionUtils.isEmpty(oldVals)) {
+            return;
+        }
 
-        // At the present moment, we are going to be filling the security.owner-producer and
-        // security.classification-system fields with the same content, the trigraph/tetragraph
-        // of the originating country/organization. That may diverge in the future
-        tempMap.put(SECURITY_OWNER_PRODUCER, this::processOwnerProducer);
-        tempMap.put(SECURITY_CLASSIFICATION_SYSTEM, this::processClassSystem);
+        Set<String> oldValues = new HashSet<>((List<String>) (List<?>) oldVals);
+        Set<String> newValues = new HashSet<>((List<String>) (List<?>) newVals);
+        if (!oldValues.equals(newValues)) {
```

Comment:
❗️ This is only going to work because the Strings have been interned, but a different JVM implementation that did not automatically intern the strings would fail. You need to assert the equality, not identity, of the elements in each set match.

There's no easy/trivial way to do that (even a library that could perform the work for you would run in `O(n<sup>2</sup>)` (fortunately, you'll be dealing with very small values). However, since you _are_ dealing with Strings, you can simplify this a bit to...

1. Create your two sets
2. Iterate set (1) and check if `set2.contains(item)`
3. Error if not

And hey...because you get that simplification, this is now reduced to `O(n)`.